src/vdc_collection.rs

```diff
@@ -1,197 +1,193 @@
-            Ok(()) => Ok(()),
-            Err(e) => Err(VdcCollectionError::StoreFailed(e)),
-        }
+        credential: Credential,
+        storage: &Arc<dyn StorageManagerInterface>,
+    ) -> Result<Key, VdcCollectionError> {
+        let val = serde_cbor::to_vec(&credential)
+            .map_err(|e| VdcCollectionError::SerializeFailed(e.to_string()))?;
+
+        storage.add(credential.as_storage_key(), Value(val))?;
+
+        // Return the key of the credential in the storage manager, in case
+        // the caller needs to cache it.
+        Ok(credential.as_storage_key())
     }
 
     /// Get a credential from the store.
-    pub fn get(&self, id: &str) -> Result<Option<Credential>, VdcCollectionError> {
-        let raw = match self.storage.get(self.str_to_key(id)) {
-            Ok(Some(x)) => x,
-            Ok(None) => return Ok(None),
-            Err(e) => return Err(VdcCollectionError::LoadFailed(e)),
-        };
-
-        match serde_cbor::de::from_slice(&raw.0) {
-            Ok(x) => Ok(x),
-            Err(_) => Err(VdcCollectionError::DeserializeFailed),
-        }
+    pub fn get(
+        &self,
+        key: Key,
+        storage: &Arc<dyn StorageManagerInterface>,
+    ) -> Result<Option<Credential>, VdcCollectionError> {
+        storage
+            .get(key)?
+            .map(|x| serde_cbor::de::from_slice(&x.0))
+            .transpose()
+            .map_err(|e| VdcCollectionError::DeserializeFailed(e.to_string()))
     }
 
     /// Remove a credential from the store.
-    pub fn delete(&self, id: &str) -> Result<(), VdcCollectionError> {
-        match self.storage.remove(self.str_to_key(id)) {
-            Ok(_) => Ok(()),
-            Err(e) => Err(VdcCollectionError::DeleteFailed(e)),
-        }
+    pub fn delete(
+        &self,
+        key: Key,
+        storage: &Arc<dyn StorageManagerInterface>,
+    ) -> Result<(), VdcCollectionError> {
+        storage.remove(key).map_err(VdcCollectionError::from)
     }
 
     /// Get a list of all the credentials.
-    pub fn all_entries(&self) -> Result<Vec<String>, VdcCollectionError> {
-        let mut r = Vec::new();
-
-        match self.storage.list() {
-            Ok(list) => {
-                for key in list {
-                    let name = key.0;
-
-                    if name.starts_with(KEY_PREFIX) {
-                        if let Some(id) = name.strip_prefix(KEY_PREFIX) {
-                            r.push(id.to_string());
-                        }
-                    }
-                }
-            }
-            Err(e) => return Err(VdcCollectionError::LoadFailed(e)),
-        }
-
-        Ok(r)
+    pub fn all_entries(
+        &self,
+        storage: &Arc<dyn StorageManagerInterface>,
+    ) -> Result<Vec<Key>, VdcCollectionError> {
+        Ok(storage
+            .list()?
+            .iter()
+            .filter(|key| key.0.contains(KEY_PREFIX))
+            .map(ToOwned::to_owned)
+            .collect())
```

Comment:
This replacement is no longer stripping `KEY_PREFIX` from the key. Is that desired? Are there other places that are relying on the prefix being stripped?